[RFC,RFT] tools: switch to new squashfs version

Ansuel opened the pull request:
Hi, this require major testing and no idea if my assumption are correct (probably not).

Anyway squashfs is now well maintained and developed so in theory we should be able to drop the now dead squashfskit4 tool and use new version of squashfs. I just ported some missing feature and now should be comparable to squashfskit4.

What i really want to understand is how to handle the lzma compression. Currently we use the old squashfs but i dropped that and I'm just compiling squashfs with lzma support (using the xz libraries) and using the -comp lzma option for these tatget...

Problem is that I don't have any affected target so I have no way to test this... 

The squashfs  patch ideally should be proposed upstream but I would first love to have some testing with all of this and also we need some size comparison of the new tool compared to using the old one.

So in short we need tons of testing with this.

Ansuel committed:
tools/squashfskit4: drop unused tool

Drop squashfskit4 tool as it got replaced by new version of squashfs4
tool.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@aparcar @ynezz @robimarko @neheb

Ansuel commented:
As always macos and their missing header :D

robimarko commented:
Nice to see this finally happening, I can test on ipq40xx and ipq807x as those are what I usually use with OpenWrt.
BTW, somebody made a PR to update squahfs-tools some time ago: https://github.com/openwrt/openwrt/pull/10743

Ansuel commented:
@robimarko totally missed it but this is a different approach to have one tool instead of having 2 that does the same thing.
2 dead tools with one maintained one looks way better than updating only squashfs

robimarko commented:
I know, I was just referring to it since it updates Squashfs-tools, similar to what you have done here.
I am obviously all for dropping squashfskit4 here.

hauke commented:
`mksquashfs-lzma` is only used in the following build commands: `append-squashfs-fakeroot-be` and `netgear-squashfs`. Both of them look like they create a fake root file system.  They probably create a fake Squashfs 3 file system to satisfy some vendor bootloader which checks if there is the correct root file system. 
The Linux kernel used in OpenWrt uses Squashfs 4 only. Squashfs 3 was never supported by mainline kernel, Squashfs 4 was added some years ago to mainline Linux.

Did you check that these commands still create a Squashfs 3 file system?

Ansuel commented:
@hauke thanks for the comments and no... i just wanted to know that! Crazy idea... Wonder if we can just spoof the version? How does these system works? The fake squash is only checked or is also used for so intermediate loading?

hauke commented:
I do not know how this works in detail. For me the code looks like we just add a empty squashfs 3 file system somewhere into the factory image.

Ansuel commented:
This thing finally compile in macos... now we need to find tester for the fake squashfs and check if they just check the version and not the entire superblock...

Ansuel commented:
btw i'm testing squashfs4 and looks to work correctly... still need some user to test the fake squashfs3

robimarko commented:
Looks like only some AVM and Netgear ath79 devices are users of old SquashFS3

aparcar commented:
Did you talk to the maintainer to see if our changes will be part of upstream?

Ansuel commented:
Still not I would like to make first some size comparison to check if I
didn't set something wrong in the compression options

Il Sab 29 Ott 2022, 10:36 Paul Spooren ***@***.***> ha
scritto:

> Did you talk to the maintainer to see if our changes will be part of
> upstream?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10967#issuecomment-1295772729>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQUMFG2FCVVF6O4WGHLWFTOXHANCNFSM6AAAAAARFLMDIA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robimarko commented:
Did you have any further luck with this?

I dont have any devices uses SquashFS3 for the fake rootfs

Ansuel commented:
AFAIK  squashfs4 works correctly... no chance to test squashfs3 tho... wonder if we should take safe approach and keep squashfs3 for now but start replacing squashfs4 with an upstream version.

robimarko commented:
Well, we just need one user with the devices using SquashFS3 to try

DragonBluep commented:
This is a full flash backup file of the Netgear WNR2200 (with squashfs3). It can be burned into any AR7241 device for testing. Hope it helps. [wnr2200bak.zip](https://github.com/openwrt/openwrt/files/10084750/wnr2200bak.zip)

Ansuel commented:
we don't need backup we need user that can test squashfs image and know how to recover from a brick.

DragonBluep commented:
We can extract the u-boot and burn it to other manufacturers' devices. If the firmware generated by the new sqashfs tool can be accepted by the WNR2200's u-boot, I think it is enough.

hauke commented on tools/Makefile:
```diff
@@ -57,7 +56,6 @@ $(curdir)/gmp/compile := $(curdir)/libtool/compile
 $(curdir)/isl/compile := $(curdir)/gmp/compile
 $(curdir)/libressl/compile := $(curdir)/pkgconf/compile
 $(curdir)/libtool/compile := $(curdir)/m4/compile $(curdir)/autoconf/compile $(curdir)/automake/compile $(curdir)/missing-macros/compile
-$(curdir)/lzma-old/compile := $(curdir)/zlib/compile
```
Is `lzma-old` still used by anything after these changes?

Ansuel replied:
think it's used for the squashfs3 that from what i can see needs to be kept while i find an user that can test my hack idea for the fake squashfs3

hauke replied:
It looks like `lzma-old` is only used by the `squashfs` application. The new one probably uses the version of the lzma library which is still developed.

dhewg commented:
> still need some user to test the fake squashfs3

The devices where this is used are:
* FRITZ!WLAN Repeater 1750E
* FRITZ!WLAN Repeater 300E
* FRITZ!Box 4020
* FRITZ!WLAN Repeater 450E
* FRITZ!WLAN Repeater DVB-C

If I read it correctly that's an empty fake squashfs? I guess that's pretty small, so you could just create that, commit the binary and use that instead?

Ansuel commented:
@dhewg needs to be created and embedded in the image afaik. it's fake in the sense that it's empty and just needed to pass uboot verification but the real image is something else.

So my idea was to have some fun with the squashfs header and fake it by adding an hack patch

dhewg commented:
Right, but if the very same thing used for all targets and won't change forever, why bother maintaining a patch?
If the target images are reproducable you can even verify it without any testers.

aparcar commented:
Hah this is terrible since we also use a tool called `fakeroot`... Maybe we should call the old fakeroot functions something like `fakefs`. @dhewg if I understand you correctly we can add a small binary blob to the repo representing a *fake squashfs* image which could be directly added instead of creating one, correct?

Ansuel commented:
We need to first test this theory.

Il Sab 4 Feb 2023, 23:10 Paul Spooren ***@***.***> ha scritto:

> Hah this is terrible since we also use a tool called fakeroot... Maybe we
> should call the old fakeroot functions something like fakefs. @dhewg
> <https://github.com/dhewg> if I understand you correctly we can add a
> small binary blob to the repo representing a *fake squashfs* image which
> could be directly added instead of creating one, correct?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10967#issuecomment-1416860943>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQXIVNU5QWXOJHQ5YDWV3HVJANCNFSM6AAAAAARFLMDIA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dhewg commented:
Hehe yeah, something like that. But that's just my understanding from briefly looking at what's blocking this PR.

But if that is indeed just a fake, empty and never changing old squashfs3 fs, and there's an issue with creating that with a different squashfs-tools version, then this may be a less painful approach.

But does anyone know why uboot on those boards even care and what they're doing with the squashfs?

aparcar commented:
Without having these devices I can't test why it's needed or if it works but here is a patch that should make things better and lose the dependency https://github.com/openwrt/openwrt/pull/11943

@hauke could you please have a look?

Spudz76 commented:
I found on eBay a grand total of one of any of the affected Fritz! models listed above.

Pretty sure literally nobody has one anymore, except the one in Greece for sale.

dhewg commented:
It's more common in the home country of fritz though: https://www.ebay-kleinanzeigen.de/s-fritz-4020/k0

Spudz76 commented:
I blame that google-shopping feature didn't bother to search eBay classifieds, then.

Figured it showed me an auction from Greece so it would show me anything from anywhere.  It's not like eBay classifieds are craigslist or something pointless to index, they are just fixed-price auctions with no end date.

dhewg commented:
Hehe, no worries.

If I read that correctly the netgear squashfs3 thing isn't just a fake/empty one as with avm, the kernel is in there.

Spudz76 commented:
Yeah I was going to say I think some devices pack the kernel inside sqfs3 for some reason (maybe the extra block-level checksums? lzma-loader wasn't invented yet? idk)

But then I wondered if my memory was playing tricks on me again, and didn't feel like investigating the actual truth, so I decided to be quiet instead of spread vague rumors.

I think the empty ones are simply there so the bootloader can "fsck" something of expected format and then continue to load, but nobody wants to actually put data in it when you could use jffs2/ubi/sqfs4 instead (once the weird bootloader expectations are faked out).  The static blob should work just fine in those.  It's like a magic-number check but a longer magic-number that is also a valid sqfs3 blob.

Ansuel commented:
Guess the squashfs3 lzma tool is there to stay... Wonder if I should just replace squashfs4 for now and handle squashfs3 later? 

Did anyone test this with squashfs4 and/or compared the size?

dhewg commented:
Didn't test v4, but had a quick look at v3. There is no -be switch on the new version, but it also doesn't seem to produce v3 compatible files even with le:
```
./mksquashfs empty empty-new.bin -comp lzma -noappend -root-owned -nopad -b 65536 -all-time 0
Parallel mksquashfs: Using 8 processors
Creating 4.0 filesystem on empty-new.bin, block size 65536.
[|                                                                                                                                  ] 0/0 100%

Exportable Squashfs 4.0 filesystem, lzma compressed, data block size 65536
	compressed data, compressed metadata, compressed fragments,
	compressed xattrs, compressed ids
	duplicates are removed
Filesystem size 0.16 Kbytes (0.00 Mbytes)
	102.53% of uncompressed filesystem size (0.15 Kbytes)
Inode table size 34 bytes (0.03 Kbytes)
	100.00% of uncompressed inode table size (34 bytes)
Directory table size 0 bytes (0.00 Kbytes)
	-nan% of uncompressed directory table size (0 bytes)
Number of duplicate files found 0
Number of inodes 1
Number of files 0
Number of fragments 0
Number of symbolic links 0
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 1
Number of hard-links 0
Number of ids (unique uids + gids) 1
Number of uids 1
	root (0)
Number of gids 1
	root (0)


./mksquashfs-lzma empty empty-old.bin -noappend -root-owned -nopad -b 65536 -fixed-time 0    
Creating little endian 3.0 filesystem on empty-old.bin, block size 65536.

Little endian filesystem, data block size 65536, compressed data, compressed metadata, compressed fragments
Filesystem size 0.15 Kbytes (0.00 Mbytes)
	99.34% of uncompressed filesystem size (0.15 Kbytes)
Inode table size 27 bytes (0.03 Kbytes)
	90.00% of uncompressed inode table size (30 bytes)
Directory table size 0 bytes (0.00 Kbytes)
	-nan% of uncompressed directory table size (0 bytes)
Number of duplicate files found 0
Number of inodes 1
Number of files 0
Number of fragments 0
Number of symbolic links  0
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 1
Number of uids 1
	root (0)
Number of gids 0

-rwx------ 1 andre andre 150 Feb  6 15:53 empty-old.bin
-rw-r----- 1 andre andre 162 Feb  6 15:53 empty-new.bin

./unsquashfs-lzma empty-new.bin 
Major/Minor mismatch, filesystem on empty-new.bin is (4:0)
I only support Squashfs 3.0 filesystems!  Later releases will support older Squashfs filesystems
```

I didn't look very hard if it somehow works, but as-is these patches won't work :\

Ansuel commented:
there is no patch that add support for v3... actually v3 and v4 header are VERY DIFFERENT... the patch here just reintroduce some additional option to tweak some compression values.

dhewg commented:
But you replaced mksquashfs-lzma with mksquashfs, that v3 to v4, no?

Ansuel commented:
tought it was just using lzma for compression... but actually the name is confusing since mksquashfs-lzma is actually squashfs3 with lzma... 

on any other image we use squashfs4 and suqashfs3 is dropped from ages now...
Also from what i remember squashfs3 was really used to fake something and I remember I didn't find any device actually putting the kernel in it... So it could be useful to parse and detail what is the situation about it.

dhewg commented:
Yeah, it's a mess.
So the remaining v3 users are the mentioned AVM and Netgear devices. The ones for AVM are empty, Netgear has a kernel in there. It looks like we need to keep the old crap.

Unless we do something like 4a0cc5d4ef29a11914850c72158aca1e2a170fb0 for those Netgear devices. If that can be solved somehow we can use #11943 for AVM and drop the two old packages.

aparcar commented:
How about proceeding with replacing squashfskit for now?

aparcar commented:
No please show some size changes and test if it still reproducible :)

Ansuel commented:
@aparcar right in theory they should produce the same image given the same compression option I will also compare that

Ansuel commented:
@aparcar Think I have good news...
From what I can see the new tool produce the same exact image.

Ansuel committed:
tools/squashfs4: add new tool for squashfs4 images

squashfs tool is finally reborn and correctly maintained.

Introduce the new version as a replacement for squasfs4kit as it was a
fork and also abandoned.

Add additional patch to add the missing feature present in squashfskit4
but still missing on this new project.

Backport each required patch that fix compilation error on macos.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
image: update LZMA_XZ_OPTIONS with new squashfs4 tool

The -Xpreset option changed format and dropped the -Xe and just require
the extreme string to be passed to the -Xpreset option.

Update the LZMA_XZ_OPTIONS to reflect this change.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Spudz76 commented:
@Ansuel here's another to maybe fix the macos problem

https://github.com/Ansuel/openwrt/pull/10

ThiloteE commented:
I own a FRITZ!WLAN Repeater 300E, but am reluctant to test anything with it. Some time ago, I tried to use this device with OpenWRT 23.02 in relayd configuration, but found that with "high" continuous troughput (34 Mbps) during iperf3 tests and single client, the repeater stalled out and lost connection. The repeater also becomes very warm during usage. After seeing those measurements, which explained why I had so crappy internet, I bought new hardware. Still keeping the FRITZ!WLAN Repeater 300E somewhere in the back of my cupboard for emergencies, but honestly, i hope I will never have to use it again...

Ansuel commented:
@Spudz76 yep dropped my patch thinking the backport patch thing was fixing the problem but NOPE... 

Also the maxos support patch won't ever get merged in this repo so we are cursed with porting them forever....

hauke commented on tools/squashfs4/Makefile:
```diff
@@ -0,0 +1,41 @@
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/plougher/squashfs-tools
+PKG_SOURCE_DATE:=2022-17-03
```
This date format is wrong. I prefer `2022-03-17`.

lovehackintosh replied:
It should be `2022-03-18`.

Spudz76 replied:
Shouldn't it be
```
PKG_SOURCE_DATE:=2023-02-08
PKG_SOURCE_VERSION:=0f5666a3b250d7732211c188b58d11b36744c75e
```

Which is the current `HEAD` over there, and the date of the commit.

Ansuel replied:
we are targeting the release hash.
@lovehackintosh why 18 ?

lovehackintosh replied:
Sorry, I got it wrong, github showed the date as 2022-03-18, but the real commit date was 2022-03-17.
![IMG_20230214_125852](https://user-images.githubusercontent.com/92633080/218643612-7a239c38-4d34-406a-98aa-b72f57c61914.jpg)
![IMG_20230214_125943](https://user-images.githubusercontent.com/92633080/218643631-7850dab8-d613-4524-a629-f82783298d6b.jpg)

Ansuel commented:
@Spudz76 can you by chance propose the change to squashfs repo ?

Ansuel commented:
I will wait for CI to finish and then I will merge this

Spudz76 commented:
PR sent upstream

https://github.com/plougher/squashfs-tools/pull/222

Spudz76 commented:
upstream PR merged already

Spudz76 commented:
Looks like that wasn't the only broken thing with macos.  I think it needs CFLAG `-std=gnu89` but may also work with `gnu99`.  Either way, C99 doesn't like sysctl calls.

Ansuel commented:
Nope problem is un the unsquashfs source that you indirectly fixed by using the correct flag and enabled the use of sysctl on that code. The include header is only for FreeBSD system but should be done also for macos

Spudz76 commented:
Well the previous macos buildbot crash said it couldn't find that header.  I don't have macos so I sort of stab in the dark.

Ansuel commented:
Well this should be the final version. If everything is ok I will merge this.

Spudz76 commented:
Yay macos worked, I guess it's cool in this case that it's the only buildbot that isn't jammed up in queue.

I see you already got it merged upstream too, huzzah!
